### PR TITLE
Front-end Challenge (Application Developer)

### PR DESCRIPTION
**Pull request by alex-kovoy**
This is a new front-end challenge for application developers.
Application developer is a fullstack engineer who specializes on user facing applications and runtimes.

### CODE REVIEW

**Comment by alex-kovoy** [comment]
@klizhentas ^^

File: challenges/frontend/challenge.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
+# Application Developer (front-end)
```
Comment (klizhentas) [inline challenges/frontend/challenge.md:1]: Frontend Application Developer

File: challenges/frontend/challenge.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+## Summary
+Implement a simple application that allows a user to browse directory content on a remote server.
```
Comment (klizhentas) [inline challenges/frontend/challenge.md:4]: a sample application?

File: challenges/frontend/challenge.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+## Rationale
+We would like to evaluate your skill in the following areas:
+- Taking existing high-level requirements and translating them to a functional application
+- Writing production level code that does not depend on gigabytes of npm packages
+- Communicating with the team when working on the challenge
+- Handling feedback from Teleport employees
```
Comment (klizhentas) [inline challenges/frontend/challenge.md:11]: Handling feedback. I would also finish every sentence with a dot here and below.

File: challenges/frontend/challenge.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+We believe this technique is not only better but also more fun compared to whiteboard/quiz interviews so common in the industry. It’s not without the downsides - it could take longer than traditional interviews. That said, it's our view that this type of challenge gives us a more accurate assessment of your ability to work well on the types of projects we’re working on day-to-day here at Teleport. [Some of the best teams use coding challenges](https://sockpuppet.org/blog/2015/03/06/the-hiring-post/). We appreciate your time and are looking forward to hacking on this project together.
+
+## Tools
+- CSS: any
+- Frontend: React, React Router
+- Backend: Nodejs|Golang|Rust
```
Comment (klizhentas) [inline challenges/frontend/challenge.md:18]: NodeJS, Golang, Rust

File: challenges/frontend/challenge.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+   - A file/folder viewer with a client-side search and sorting capabilities (level 1-2)
+   - Authentication with a login screen and backend API to browse a directory  (levels 2-4)
+
+  There are 6 engineering levels at Teleport. It’s possible to score on levels 1-4 through this coding challenge; each level has its own set of requirements.
+
+### Level 1-2
+  This level focuses on JS/CSS/DOM knowledge and does not require any server-side logic to implement.
+
+  Create a directory viewer with the following functionality:
+  - An ability to view and browse directory content
```
Comment (klizhentas) [inline challenges/frontend/challenge.md:34]: I'd remove 'An ability'.

File: challenges/frontend/challenge.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+      name: "README.md",
+      sizeKb: 4340,
+      type: "file",
+    },
+  ],
+};
+```
+
+  - You can use a design of similar viewers like Google Drive or Github
+    as a starting point and then add missing pieces to it. For example, you can borrow css from these design systems:
```
Comment (klizhentas) [inline challenges/frontend/challenge.md:78]: CSS

File: challenges/frontend/challenge.md [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+- Avoid using unnecessary 3d party dependencies.
+
+- Error handling. We pay extra attention to error handling. Make sure that they are properly handled and not ignored.
+
+- URL navigation is a big part of great UX, do not ignore it.
+
+- Keep your CSS simple but not simpler. Do not waste your time on animations, instead make sure that an extra character does not destroy your layout.
+
+- Make sure that your code is secured and your application is not vulnerable to CSRF/XSS attacks.
```
Comment (klizhentas) [inline challenges/frontend/challenge.md:144]: I'd also add a bit on secure session management and good crypto setup for server side for senior devs.

**Review by klizhentas — APPROVED** [review]

**Pull request merged**